### PR TITLE
Digital Ocean hosting guide

### PR DESCRIPTION
**Pull request by ChrisChinchilla**
A lot of this uses content from the Socket Compose guide, which I will back-ort, but might be worth using content reuse somehow.

### CODE REVIEW

**Comment by StarfallProjects** [comment]
@ChrisChinchilla a few things to address:

- UI elements should be bold (e.g. "Click **Droplets** from the **Manage** menu)
- Please ensure external links open in a new tab using the syntax described [here](https://github.com/n8n-io/n8n-docs/wiki/Styles#links)
- I'd suggest following the convention around nouns/verbs and one word/two word. So 'Login to Droplet' would be 'Log in to Droplet', 'Setup DNS' would be 'Set up DNS', and so on.
- While it's reasonable to assume some existing knowledge with these guides, I think it would be good to offer a little more handholding. For instance, when telling users to create files somewhere accessible by Docker, can we either give examples, or links to docs that will give them a clue? Could we link to guides on configuring DNS to point to a droplet? And so on.

**Comment by ChrisChinchilla** [comment]
> While it's reasonable to assume some existing knowledge with these guides, I think it would be good to offer a little more handholding. For instance, when telling users to create files somewhere accessible by Docker, can we either give examples, or links to docs that will give them a clue? Could we link to guides on configuring DNS to point to a droplet? And so on.

This one is a bit tricky… The DNS especially as it massively depends on who hosts their DNS, and there's hundreds, if not thousands of companies that might do that, and they all do it in slight different ways, but can probably link to general DNS guides, if they exist somewhere. For example, most of the DNS guides on DO only cover how to use _their_ DNS service… The Docker one I can add though, if I can find something good and useful.

**Comment by ChrisChinchilla** [comment]
@StarfallProjects OK done… For the rootless/docker user one, I added something, but to be honest, if someone is creating linux users, we can probably assume they do know what they're doing and mostly just need reminding in a Docker context.

File: docs/hosting/server-setups/digital-ocean.md [hunk 1]
```diff
@@ -0,0 +1,178 @@
+# Hosting n8n on DigitalOcean
+
+This hosting guide shows you how to self-host n8n on a DigitalOcean droplet. It uses n8n and [Caddy](http://caddyserver.com){:target="_blank"} to allow access to the Droplet from the internet (a reverse proxy). It uses [Docker Compose](https://docs.docker.com/compose/){:target="_blank"} to create define the application components and how they work together.
```
Comment (StarfallProjects) [inline docs/hosting/server-setups/digital-ocean.md:3]:
```suggestion
This hosting guide shows you how to self-host n8n on a DigitalOcean droplet. It uses n8n and [Caddy](http://caddyserver.com){:target="_blank" class=.external-link} (a reverse proxy) to allow access to the Droplet from the internet . It uses [Docker Compose](https://docs.docker.com/compose/){:target="_blank"} to create and define the application components and how they work together.
```

File: docs/hosting/server-setups/digital-ocean.md [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+## Create droplet
```
Comment (StarfallProjects) [inline docs/hosting/server-setups/digital-ocean.md:5]:
```suggestion
## Create a Droplet
```

File: docs/hosting/server-setups/digital-ocean.md [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+Select the project to host the Droplet and click __Droplets__ from the __Manage__ menu. You can use [the Docker image](https://marketplace.digitalocean.com/apps/docker){:target="_blank"} from the __Marketplace__ tab or use one of the Linux distributions and [install Docker yourself](https://www.docker.com/get-started/){:target="_blank"}. The rest of this tutorial assumes you have Docker installed on the Droplet.
```
Comment (StarfallProjects) [inline docs/hosting/server-setups/digital-ocean.md:7]:
```suggestion
Select the project to host the Droplet and click **Droplets** from the **Manage** menu. You can use [the Docker image](https://marketplace.digitalocean.com/apps/docker){:target="_blank" class=.external-link} from the **Marketplace** tab or use one of the Linux distributions and [install Docker yourself](https://www.docker.com/get-started/){:target="_blank" class=.external-link}. The rest of this tutorial assumes you have Docker installed on the Droplet.
```

Comment (StarfallProjects) [inline docs/hosting/server-setups/digital-ocean.md:7]: Where are they doing this? Maybe worth putting them in context. Something like "Log in to DigitalOcean. From your dashboard/homepage (?) . . . "

File: docs/hosting/server-setups/digital-ocean.md [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+
+## Login to Droplet
```
Comment (StarfallProjects) [inline docs/hosting/server-setups/digital-ocean.md:9]:
```suggestion
## Log in to your Droplet
```

File: docs/hosting/server-setups/digital-ocean.md [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+
+The remainder of the steps in this guide require you to log in to the Droplet via a terminal with SSH. [Find more details on how to do this in the Digital Ocean documentation](https://docs.digitalocean.com/products/droplets/how-to/connect-with-ssh/){:target="_blank"}.
```
Comment (StarfallProjects) [inline docs/hosting/server-setups/digital-ocean.md:11]:
```suggestion
The remainder of the steps in this guide require you to log in to the Droplet using a terminal with SSH. [Find more details on how to do this in the Digital Ocean documentation](https://docs.digitalocean.com/products/droplets/how-to/connect-with-ssh/){:target="_blank"}.
```

File: docs/hosting/server-setups/digital-ocean.md [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+
+## Create folders and files
+
+Both n8n and Caddy require creating folders that the host operating system (the Digital Ocean Droplet) copies to Docker containers to make them available to Docker.
+
+Create the following on the Droplet in a location accessible by Docker. If you run Docker as root user, you can create them anywhere. If you run Docker "[Rootless](https://docs.docker.com/engine/security/rootless/){:target="_blank"}" for better security, create them in a location that user has access to:
+
+- _caddy_config_: Holds the Caddy configuration files.
```
Comment (StarfallProjects) [inline docs/hosting/server-setups/digital-ocean.md:19]:
```suggestion
- `caddy_config`: Holds the Caddy configuration files.
```

File: docs/hosting/server-setups/digital-ocean.md [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+- _caddy_data_: A cache folder for Caddy.
```
Comment (StarfallProjects) [inline docs/hosting/server-setups/digital-ocean.md:20]:
```suggestion
- `caddy_data`: A cache folder for Caddy.
```

File: docs/hosting/server-setups/digital-ocean.md [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+- _local_files_: A folder for files you upload or add via n8n.
```
Comment (StarfallProjects) [inline docs/hosting/server-setups/digital-ocean.md:21]:
```suggestion
- `local_files`: A folder for files you upload or add using n8n.
```

File: docs/hosting/server-setups/digital-ocean.md [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+
+Select the project to host the Droplet and click __Droplets__ from the __Manage__ menu. You can use [the Docker image](https://marketplace.digitalocean.com/apps/docker){:target="_blank"} from the __Marketplace__ tab or use one of the Linux distributions and [install Docker yourself](https://www.docker.com/get-started/){:target="_blank"}. The rest of this tutorial assumes you have Docker installed on the Droplet.
+
+## Login to Droplet
+
+The remainder of the steps in this guide require you to log in to the Droplet via a terminal with SSH. [Find more details on how to do this in the Digital Ocean documentation](https://docs.digitalocean.com/products/droplets/how-to/connect-with-ssh/){:target="_blank"}.
+
+## Create folders and files
+
+Both n8n and Caddy require creating folders that the host operating system (the Digital Ocean Droplet) copies to Docker containers to make them available to Docker.
```
Comment (StarfallProjects) [inline docs/hosting/server-setups/digital-ocean.md:15]:
```suggestion
Both n8n and Caddy require creating folders that the host operating system (the DigitalOcean Droplet) copies to Docker containers to make them available to Docker.
```

File: docs/hosting/server-setups/digital-ocean.md [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+
+### Create Docker volume
+
+To persist the Caddy cache between restarts and speed up start times, create [a Docker volume](https://docs.docker.com/storage/volumes/){:target="_blank"} that Docker reuses between restarts.
+
+```shell
+docker volume create caddy_data
+```
+
+## Setup DNS
```
Comment (StarfallProjects) [inline docs/hosting/server-setups/digital-ocean.md:37]:
```suggestion
## Set up DNS
```

File: docs/hosting/server-setups/digital-ocean.md [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+Open the following ports in the Droplet's firewall by running the following two commands:
+
+```shell
+sudo ufw allow 80
+sudo ufw allow 443
+```
+
+## Configure n8n
+
+Create an _.env_ file in the same folder you will run Docker Compose from, and add the following, replacing the values with your own:
```
Comment (StarfallProjects) [inline docs/hosting/server-setups/digital-ocean.md:54]:
```suggestion
Create a `.env` file in the same folder you run Docker Compose from, and add the following, replacing the values with your own:
```

File: docs/hosting/server-setups/digital-ocean.md [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+# If not set New York time will be used
+GENERIC_TIMEZONE=Europe/Berlin
+
+# The email address to use for the SSL certificate creation
+SSL_EMAIL=example@example.com
+```
+
+## Create Docker Compose file
+
+The Docker compose file defines the services the application needs, in this case Caddy and n8n.
```
Comment (StarfallProjects) [inline docs/hosting/server-setups/digital-ocean.md:93]:
```suggestion
The Docker Compose file defines the services the application needs, in this case Caddy and n8n.
```

File: docs/hosting/server-setups/digital-ocean.md [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+
+- The Caddy service definition defines the ports it uses and the local volumes to copy to the containers.
+- The n8n service definition defines the ports it uses, the environment variables n8n needs to run (some defined in the _.env_ file), and the volumes it needs to copy to the containers.
+
+Create a _docker-compose.yml_ file, make sure to create it in the same location as the _.env_ file:
```
Comment (StarfallProjects) [inline docs/hosting/server-setups/digital-ocean.md:98]:
```suggestion
Create a `docker-compose.yml` file. Make sure to create it in the same location as the `.env` file:
```

File: docs/hosting/server-setups/digital-ocean.md [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+# The email address to use for the SSL certificate creation
+SSL_EMAIL=example@example.com
+```
+
+## Create Docker Compose file
+
+The Docker compose file defines the services the application needs, in this case Caddy and n8n.
+
+- The Caddy service definition defines the ports it uses and the local volumes to copy to the containers.
+- The n8n service definition defines the ports it uses, the environment variables n8n needs to run (some defined in the _.env_ file), and the volumes it needs to copy to the containers.
```
Comment (StarfallProjects) [inline docs/hosting/server-setups/digital-ocean.md:96]:
```suggestion
- The n8n service definition defines the ports it uses, the environment variables n8n needs to run (some defined in the `.env` file), and the volumes it needs to copy to the containers.
```

File: docs/hosting/server-setups/digital-ocean.md [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+
+volumes:
+  caddy_data:
+    external: true
+  caddy_config:
+```
+
+## Configure Caddy
+
+Caddy needs to know which domains it should serve, and which port to expose to the outside world. Create a _Caddyfile_ file in the *caddy_config* folder.
```
Comment (StarfallProjects) [inline docs/hosting/server-setups/digital-ocean.md:148]:
```suggestion
Caddy needs to know which domains it should serve, and which port to expose to the outside world. Create a `Caddyfile` file in the `caddy_config` folder.
```

File: docs/hosting/server-setups/digital-ocean.md [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+
+```shell
+nano caddy_config/Caddyfile
+```
+
+Add the following configuration, adding your subdomain. The `n8n` tells Caddy to use the service definition defined in the _docker-compose.yml_ file:
```
Comment (StarfallProjects) [inline docs/hosting/server-setups/digital-ocean.md:154]:
```suggestion
Add the following configuration, adding your subdomain. The `n8n` tells Caddy to use the service definition defined in the `docker-compose.yml` file:
```

File: docs/hosting/server-setups/digital-ocean.md [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+
+Start n8n and Caddy with the following command:
+
+```shell
+docker-compose up -d
+```
+
+Open the URL formed of the subdomain and domain name defined earlier, enter the user name and password defined earlier, and you should be able to access n8n.
+
+Stop n8n and Caddy with the following command:
```
Comment (StarfallProjects) [inline docs/hosting/server-setups/digital-ocean.md:174]:
```suggestion
You can stop n8n and Caddy with the following command:
```

Comment (StarfallProjects) [reply docs/hosting/server-setups/digital-ocean.md:174]: I don't want to sound like we're ordering them to stop their instance. This makes it clearer it's an option I think.

**Review by StarfallProjects — CHANGES_REQUESTED** [review]
Mostly formatting, some of which is fairly recently defined (sorry!)

File: docs/hosting/server-setups/digital-ocean.md [hunk 18]
```diff
@@ -0,0 +1,178 @@
...
+Create the following on the Droplet in your home folder:
+
+- `caddy_config`: Holds the Caddy configuration files.
+- `caddy_data`: A cache folder for Caddy.
+- `local_files`: A folder for files you upload or add via n8n.
+
+Use the following command to create the folders:
+
+```shell
+mkdir caddy_config caddy_data local_files
```
Comment (StarfallProjects) [inline docs/hosting/server-setups/digital-ocean.md:26]: Should we also tell them to create an n8n directory? The subsequent .env file expects everything to be in root/n8n, not just home dir

File: docs/hosting/server-setups/digital-ocean.md [hunk 19]
```diff
@@ -0,0 +1,178 @@
...
+```
+
+### Create Docker volume
+
+To persist the Caddy cache between restarts and speed up start times, create [a Docker volume](https://docs.docker.com/storage/volumes/){:target="_blank"} that Docker reuses between restarts.
+
+```shell
+docker volume create caddy_data
```
Comment (StarfallProjects) [inline docs/hosting/server-setups/digital-ocean.md:34]: And depending on answer to the above, should this also be in the n8n directory?

File: docs/hosting/server-setups/digital-ocean.md [hunk 20]
```diff
@@ -0,0 +1,178 @@
...
+Open the following ports in the Droplet's firewall by running the following two commands:
+
+```shell
+sudo ufw allow 80
+sudo ufw allow 443
+```
+
+## Configure n8n
+
+Create an `.env` file in the same folder you will run Docker Compose from, and add the following, replacing the values with your own:
```
Comment (StarfallProjects) [inline docs/hosting/server-setups/digital-ocean.md:54]: Have we talked about this before? Might be friendly to specify a directory?

Comment (ChrisChinchilla) [reply docs/hosting/server-setups/digital-ocean.md:54]: If we specify a directory, we'd need to update a lot of other paths in the guide to reflect that. Typically they're found in a project folder or the home folder, which we are using in this case. Personally I'd say keeping the folder structure simpler is possibly more friendly as then there's less to do.

Not quite sure what you mean by "have we talked about this before" though? It's in the current Docker compose tutorial, which is where I took it from. Ideally you don't want to commit sensitive information to a repository and outside of something like Kubernetes, which has an entire mechanism for this, a standard method is to use environment variables.

**Pull request closed without merging**